src/psyclone/psyir/transformations/fold_conditional_return_expressions_trans.py

```diff
@@ -0,0 +1,148 @@
+#   this software without specific prior written permission.
+#
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+# "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+# LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+# FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+# COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+# INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+# BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+# LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+# LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+# ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+# -----------------------------------------------------------------------------
+# Author S. Siso, STFC Daresbury Lab
+
+'''This module contains the FoldConditionalReturnExpressionsTrans. '''
+
+from psyclone.psyir.nodes import (Routine, IfBlock, Return,
+                                  UnaryOperation)
+from psyclone.psyGen import Transformation
+from psyclone.psyir.transformations.transformation_error import \
+        TransformationError
+
+
+class FoldConditionalReturnExpressionsTrans(Transformation):
+    ''' Provides a transformation that folds conditional expressions with only
+    a return statement inside so that the Return statement can disapear
+    (because happens at the end of the Routine). For example, the following
+    code:
+
+    >>> subroutine test(i)
+    >>>    if (i < 5) then
+    >>>        return
+    >>>    endif
+    >>>    if (i > 10) then
+    >>>        return
+    >>>    endif
+    >>>    ! CODE
+    >>>  end subroutine
+
+    will be transformed to:
+
+    >>> subroutine test(i)
+    >>>    if (.not.(i < 5)) then
+    >>>        if (.not.(i > 10)) then
+    >>>            ! CODE
+    >>>        endif
+    >>>    endif
+    >>>  end subroutine
+
+    '''
+
+    def __str__(self):
+        return "Fold all conditional expressions with Return statements."
+
+    @property
+    def name(self):
+        '''Returns the name of this transformation as a string.'''
+        return "FoldConditionalReturnExpressionsTrans"
+
+    def validate(self, node, options=None):
+        '''Ensure that it is valid to apply this transformation to the
+        supplied node.
+
+        :param node: the node to validate.
+        :type node: :py:class:`psyclone.psyir.nodes.Routine`
+        :param options: a dictionary with options for transformations.
+        :type options: dictionary of string:values or None
+
+        :raises TransformationError: if the node is not a Routine.
+        '''
+
+        if not isinstance(node, Routine):
+            raise TransformationError("Error in {0} transformation. "
+                                      "This transformation can only be applied"
```

Comment:
Please include the type of the supplied node in the message.

Reply:
Done